assets/src/edit-story/components/carousel/carouselProvider.js

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { useCallback, useLayoutEffect, useRef, useMemo, useState } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '../../app';
+import CarouselContext from './carouselContext';
+import useCarouselSizing from './useCarouselSizing';
+import useCarouselScroll from './useCarouselScroll';
+import useCarouselKeys from './useCarouselKeys';
+
+function CarouselProvider({ availableSpace, children }) {
+  const {
+    pages,
+    currentPageId,
+    setCurrentPage,
+    arrangePage,
+  } = useStory(
+    ({
+      state: { pages, currentPageId },
+      actions: { setCurrentPage, arrangePage },
+    }) => ({ pages, currentPageId, setCurrentPage, arrangePage })
+  );
+
+  const [listElement, setListElement] = useState(null);
+  const pageRefs = useRef([]);
+
+  const numPages = pages.length;
+
+  const pageIds = useMemo(() => pages.map(({ id }) => id), [pages]);
+
+  const {
+    pageThumbWidth,
+    pageThumbHeight,
+    pageThumbMargin,
+    carouselWidth,
+    hasOverflow,
+  } = useCarouselSizing({ availableSpace, numPages });
+
+  const setPageRef = useCallback((page, el) => {
+    pageRefs.current[page.id] = el;
+  }, []);
+
+  const {
+    canScrollBack,
+    canScrollForward,
+    scrollBack,
+    scrollForward,
+  } = useCarouselScroll({ listElement, carouselWidth, hasOverflow });
+
+  useCarouselKeys({ listElement, pageRefs });
+
+  useLayoutEffect(() => {
+    if (!hasOverflow) {
+      return;
+    }
+    const currentPageRef = pageRefs.current[currentPageId];
+
+    if (!currentPageRef || !currentPageRef.scrollIntoView) {
+      return;
+    }
+
+    currentPageRef.scrollIntoView({
+      inline: 'center',
+      behavior: 'smooth',
+    });
```

Comment:
While we're at it, might be a good time to address #5077 by using `scrollBy` with an offset instead of `scrollIntoView`.

Is that easily possible?

Reply:
Hm, it should be. I should be able to calculate all the numbers quite easily.

Reply:
Done in 72843dd80 - and it looks freaking nice, if I can say so myself. I actually [ended up using `scrollTo` rather than `scrollBy`](https://github.com/google/web-stories-wp/pull/5941/files#diff-60efa86d6d8735e06b04393cb877f4fa0242b75724db8c071b4d135da09b2692R100-R153) though 🤷